Documentation for Flow Exporter feature (#980)
vmware-tanzu/antrea: docs/network-flow-visibility.md

    # the flow exporter that sends IPFIX flow records of conntrack flows on OVS bridge. If no L4 transport proto is given,
    # we consider tcp as default.
    flowCollectorAddr: "192.168.86.86:4739:tcp"

    # Provide flow poll interval as a duration string. This determines how often the flow exporter dumps connections from the conntrack module.
    # Flow poll interval should be greater than or equal to 1s (one second).
    # Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".
    flowPollInterval: "1s"

    # Provide flow export frequency, which is the number of poll cycles elapsed before flow exporter exports flow records to
    # the flow collector.
    # Flow export frequency should be greater than or equal to 1.
    flowExportFrequency: 5
```
 
Please note that the default values for `flowPollInterval` and `flowExportFrequency` parameters are set to 5s and 12, respectively.
`flowCollectorAddr` is a required parameter that is necessary for the Flow Exporter feature to work.

### Supported capabilities

Currently, the Flow Exporter feature provides visibility for Pod-to-Pod, Pod-to-Node, Node-to-Pod, Node-to-Node and Pod-to-Service

> network flows along with the associated statistics such as byte throughput, packet throughput, cumulative byte count, cumulative packet

Reviewer: "byte throughput, packet throughput" 
I understand the meanings given the context. Are these commonly used terms though? I do see we have "Node Throughput" below and these two could confuse with it.

Author: Hi Kai,
I see that packet throughput to be widely used, however byte throughput may be considered informal. Changed it to data throughput and also added units to avoid confusion with metrics shown under Node Throughput.

> # Network Flow Visibility in Antrea

Reviewer 2: may i also suggest adding a "table of content" for easier navigation here..

[Antrea](architecture.md) is a Kubernetes network plugin that provides network connectivity and security features for Pod workloads.
Considering the scale and dynamism of Kubernetes workloads in a cluster, Network Flow Visibility helps in the management

> and configuration of Kubernetes constructs such as Network Policy, Services, Pods etc., and and thereby provides opportunities

Reviewer 2: ```suggestion
and configuration of Kubernetes resources such as Network Policy, Services, Pods etc., and thereby provides opportunities
```

to enhance the performance and security aspects of Pod workloads.

For visualizing the network flows, Antrea monitors the flows in Linux conntrack module. These flows are converted to flow records
and are sent to the configured flow controller. High-level design is given below:

<img src="https://s3-us-west-2.amazonaws.com/downloads.antrea.io/static/flow_visibility.png" width="600" alt="Network Flow Visibilty">

## Flow Exporter feature

In Antrea, the basic building block for the Network Flow Visibility is the **Flow Exporter feature**. Flow Exporter operates

> at Antrea Agent; it builds and maintains a connection store by polling and dumping flows from conntrack module periodically.

Reviewer 2: ```suggestion
within the Antrea Agent; it builds and maintains a connection store by polling and dumping flows from conntrack module periodically.
```
?

    # FeatureGates is a map of feature names to bools that enable or disable experimental features.
    featureGates:
    # Enable flowexporter which exports polled conntrack connections as IPFIX flow records from each agent to a configured collector.
      FlowExporter: true

    # Provide flow collector address as string with format <IP>:<port>[:<proto>], where proto is tcp or udp. This also enables
    # the flow exporter that sends IPFIX flow records of conntrack flows on OVS bridge. If no L4 transport proto is given,
    # we consider tcp as default.
    flowCollectorAddr: "192.168.86.86:4739:tcp"

    # Provide flow poll interval as a duration string. This determines how often the flow exporter dumps connections from the conntrack module.
    # Flow poll interval should be greater than or equal to 1s (one second).
    # Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".
    flowPollInterval: "1s"

    # Provide flow export frequency, which is the number of poll cycles elapsed before flow exporter exports flow records to
    # the flow collector.
    # Flow export frequency should be greater than or equal to 1.
    flowExportFrequency: 5
``` 

> Please note that if the flowPollInterval and flowExportFrequency parameters are not set with the Flow Exporter feature enabled,

Reviewer 2: ```suggestion
Please note that the default values for `flowPollInterval` and `flowExportFrequency` parameters are set to 5s and 12, respectively. `flowCollectorAddr` is a required parameter necessary for Flow Exporter feature to work.
```

then they take the default values of 5s and 12 respectively. However, the flowCollectorAddr parameter must be provided for the Flow Exporter
feature to work. 

### Supported capabilities

Currently, the Flow Exporter feature provides visibility to Pod-to-Pod, Pod-to-Node, Node-to-Pod, Node-to-Node and Pod-to-Service
network flows along with the associated statistics such as byte throughput, packet throughput, cumulative byte count, cumulative packet
count etc. Pod-To-Service flow visibility is supported only when Antrea Proxy enabled. 

> Kubernetes info such as Node name, Pod name, Pod namespace, Service name etc. are added to the flow records.

Reviewer 2: ```suggestion
Kubernetes info such as Node name, Pod name, Pod Namespace, Service name etc. are added to the flow records.
```

# Network Flow Visibility in Antrea

[Antrea](architecture.md) implements the Container Network Interface(CNI) and Network Policy (both Kubernetes and custom)
to provide network connectivity and security features for pod workloads. Considering the scale and dynamism of Kubernetes
workloads in a cluster, Network Flow Visibility helps in the management and configuration of Kubernetes constructs such as Network
Policy, Services, Pods etc., and and thereby provides opportunities to enhance the performance and security aspects of Pod workloads.

For visualizing the network flows, Antrea monitors the flows in conntrack module. These flows are converted to flow records
and are sent to the configured flow controller. High-level design is given below:

> <img src="https://s3-us-west-2.amazonaws.com/downloads.antrea.io/static/flow_visibility.png" width="600" alt="Network Flow Visibilty">

Author: Yes, I revised the picture. Contacted Zhengsheng to upload the latest one.

    flowCollectorAddr: "192.168.86.86:4739:tcp"

    # Provide flow poll interval as a duration string. This determines how often the flow exporter dumps connections from the conntrack module.
    # Flow poll interval should be greater than or equal to 1s (one second).
    # Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".
    flowPollInterval: "1s"

    # Provide flow export frequency, which is the number of poll cycles elapsed before flow exporter exports flow records to
    # the flow collector.
    # Flow export frequency should be greater than or equal to 1.
    flowExportFrequency: 5
``` 

Please note that if the flowPollInterval and flowExportFrequency parameters are not set with the Flow Exporter feature enabled,
then they take the default values of 5s and 12 respectively. However, the flowCollectorAddr parameter must be provided for the Flow Exporter
feature to work. 

### Supported capabilities

Currently, the Flow Exporter feature provides visibility to Pod-to-Pod, Pod-to-Node, Node-to-Pod, Node-to-Node and Pod-to-Service
network flows along with the associated statistics such as byte throughput, packet throughput, cumulative byte count, cumulative packet

> count etc. Pod-To-Service flow visibility is supported only when Antrea Proxy enabled. In the future, this support will

Reviewer 3: I am not sure if we want to support kube-proxy for long. Consider removing the last sentence.

Author: Sure, Jianjun removed it. A follow-up question- Even though kube proxy is not going to be the default, will there be no option exposed to turn it on?

Reviewer 3: I think we have not concluded this yet. But I personally hope Antrea Proxy will be the default and probably for the majority of clusters.

be extended to kube-proxy. 

Kubernetes info such as Node name, Pod name, Pod namespace, Service name etc. are added to the flow records.
For flow records that are exported from any given Antrea Agent, we provide only the information of Kubernetes entities
that are local to the Antrea Agent. In the future, we plan to extend this feature to provide remote Kubernetes entities
such as remote Node name, remote Pod Name etc.

> Please note that in the case of inter-node flows, we are exporting only one copy of the flow record from the source host,

Reviewer 3: node -> Node
source host -> source Node

> where the flow is originated from and ignore the flow record from the destination host, where the destination pod resides.

Reviewer 3: host -> Node
pod -> Pod

> In the future, this behavior will be changed when the support for network policy comes in as both hosts may apply different

Reviewer 3: Network Policy

> network policies and rules.

Reviewer 3: Network Policies

  * [Supported capabilities](#supported-capabilities)
- [ELK Flow Collector](#elk-flow-collector)
  * [Purpose](#purpose)
  * [About Elastic Stack](#about-elastic-stack)
  * [Deployment Steps](#deployment-steps)
  * [Pre-built Dashboards](#pre-built-dashboards)
    + [Overview](#overview-1)
    + [Flows](#flows)
      - [Pod-to-Pod Traffic](#pod-to-pod-traffic)
      - [Pod-to-Service Traffic](#pod-to-service-traffic)
    + [Flow Records](#flow-records)
    + [Node Throughput](#node-throughput)
      
## Overview
[Antrea](architecture.md) is a Kubernetes network plugin that provides network connectivity and security features for Pod workloads.
Considering the scale and dynamism of Kubernetes workloads in a cluster, Network Flow Visibility helps in the management
and configuration of Kubernetes resources such as Network Policy, Services, Pods etc., and thereby provides opportunities
to enhance the performance and security aspects of Pod workloads.

For visualizing the network flows, Antrea monitors the flows in Linux conntrack module. These flows are converted to flow records
and are sent to the configured flow controller. High-level design is given below:

> <img src="https://s3-us-west-2.amazonaws.com/downloads.antrea.io/static/netviz.png" width="600" alt="Network Flow Visibilty">

Reviewer 4: I think we should consider building a "cleaner" svg diagram in the future, that we can potentially use for slide presentations. But this is good enough got now.

Author: Yes, will keep a note of that.

## Flow Exporter feature

In Antrea, the basic building block for the Network Flow Visibility is the **Flow Exporter feature**. Flow Exporter operates
within Antrea Agent; it builds and maintains a connection store by polling and dumping flows from conntrack module periodically.

> The connections from connection store are exported to a flow collector using IPFIX protocol, and for this purpose we

Reviewer 4: s/The connections from connection store are exported to a flow collector using IPFIX protocol/Connections from the connection store are exported to a flow collector using the IPFIX protocol

> use [go-ipfix](https://github.com/vmware/go-ipfix) library.

Reviewer 4: use *the* go-ipfix library

 
### Configuration

To enable the Flow Exporter feature at the Antrea Agent, the following config parameters have to be set in the Antrea Agent
ConfigMap as shown below. We provide some examples for the parameter values in the following snippet.

```yaml

>   antrea-agent.conf: |

Reviewer 4: Consider showing the `AntreaProxy` feature set to true here as well since it unlocks more features

      FlowExporter: true

    # Provide flow collector address as string with format <IP>:<port>[:<proto>], where proto is tcp or udp. This also enables
    # the flow exporter that sends IPFIX flow records of conntrack flows on OVS bridge. If no L4 transport proto is given,
    # we consider tcp as default.
    flowCollectorAddr: "192.168.86.86:4739:tcp"

    # Provide flow poll interval as a duration string. This determines how often the flow exporter dumps connections from the conntrack module.
    # Flow poll interval should be greater than or equal to 1s (one second).
    # Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".
    flowPollInterval: "1s"

    # Provide flow export frequency, which is the number of poll cycles elapsed before flow exporter exports flow records to
    # the flow collector.
    # Flow export frequency should be greater than or equal to 1.
    flowExportFrequency: 5
```
 
Please note that the default values for `flowPollInterval` and `flowExportFrequency` parameters are set to 5s and 12, respectively.
`flowCollectorAddr` is a required parameter that is necessary for the Flow Exporter feature to work.

### Supported capabilities

> Currently, the Flow Exporter feature provides visibility to Pod-to-Pod, Pod-to-Node, Node-to-Pod, Node-to-Node and Pod-to-Service

Reviewer 4: s/visibility to/visibility for

network flows along with the associated statistics such as byte throughput, packet throughput, cumulative byte count, cumulative packet

> count etc. Pod-To-Service flow visibility is supported only when Antrea Proxy enabled. 

Reviewer 4: "when Antrea Proxy enabled" --- turn this into a markdown link to the the features gate doc

> Kubernetes info such as Node name, Pod name, Pod Namespace, Service name etc. are added to the flow records.

Reviewer 4: information is an uncountable noun so the verb should be singular (is added)

Reviewer 4: also use "information" instead of "info" for consistency with the rest of the doc

> For flow records that are exported from any given Antrea Agent, we provide only the information of Kubernetes entities

Reviewer 4: s/we provide only the information of Kubernetes entities/we only provide the Kubernetes information for entities

> that are local to the Antrea Agent. In the future, we plan to extend this feature to provide remote Kubernetes entities

Reviewer 4: s/to provide remote Kubernetes entities/to provide information about remote entities

such as remote Node name, remote Pod name etc.

> Please note that in the case of inter-node flows, we are exporting only one copy of the flow record from the source Node,

Reviewer 4: s/node/Node

Author: Prefer to keep it as inter-node as node here is about any general node and not a particular one to make it proper noun. What do you think?

Reviewer 4: Yes we always capitalize it by convention, it's not related to the semantics of the sentence

> where the flow is originated from and ignore the flow record from the destination Node, where the destination Pod resides.

Reviewer 4: s/from and ignore/from, and ignore

> In the future, this behavior will be changed when the support for Network Policy comes in as both hosts may apply different

Reviewer 4: s/comes in/is added

Network Policies and Rules.

## ELK Flow Collector

### Purpose

> Antrea supports sending IPFIX flow records through the Flow Exporter feature explained above. The Elastic

Reviewer 4: s/explained/described

[Kibana](https://www.elastic.co/kibana/) is mainly for data visualization and
exploration.

### Deployment Steps

> To create all the necessary resources in the `elk-flow-collector` namespace

Reviewer 2: ```suggestion
To create all the necessary resources in the `elk-flow-collector` Namespace
```
